design: Update GatewayAPI design doc (#3797)
projectcontour/contour: design/gateway-apis-implementation.md

### TLSRoute

Contour will watch all TLSRoute objects, and filter for entries matching a label selector in its configured Gateway's spec.listeners[].routes, and will configure the associated routes.
Configuration of TLSRoutes will be subject to the rules around the RouteGateways field.

Contour will only ever update the status of TLSRoute objects.

Errors or conflicts here will render that section of the config invalid. Other valid sections will still be passed to Envoy.

The output of this watcher is Envoy configuration.

## Detailed Design

### Responsibilities

#### Contour

In the fully managed mode, Contour is responsible for managing the Envoy fleet which includes the Envoy Kubernetes Daemonset and the Envoy Kubernetes Service.

#### Operator

The operator is responsible for managing instances of Contour as well as any additional configuration items that are related to running Contour or Envoy.
These pieces will be configured in the Contour Configuration CRD.

### Configmap to CRD

Currently, Contour gets its configuration from two different places, one is the configuration file represented as a Kubernetes configmap.
The other are flags which are passed to Contour.

> This design will migrate the Contour configuration configmap to a full custom resource definition which will allow for better validation, easier processing, as well as represent a full Contour configuration including how Envoy is to be provisioned if the user desires.

Reviewer: Are we planning to continue supporting the config file as well until Contour 2.0? Otherwise seems like this would constitute a breaking change.

Author: I think this would only support the crd. So we'd deprecate support for managed bits using a configmap.

Reviewer 2: I agree, I think the stance here is:
- Contour proper continues supporting the config file until we release a 2.0 (which may be after October), but it's deprecated.
- The fully-managed, operator flow requires the CRD before then.

#### Benefits

* Eliminates the need to translate from a CRD to a configmap (Like the Operator does today)
* Allows for a place to surface information about configuration errors - the CRD status, in addition to the Contour log files

> * Allows the Operator workflow to match a non-operator workflow  (i.e. you start with a Contour configuration CRD)

Reviewer: What exactly do you mean by this? Are you implying that [the Operator's Contour CRD](https://github.com/projectcontour/contour-operator/blob/main/api/v1alpha1/contour_types.go#L38-L50) goes away and is replaced by the new Contour config CRD? If so, I'm not clear where an operator user would configure things like the namespace to install Contour in, or the # of replicas in the Contour deployment, etc. But if that's not what you meant, disregard.

Author: Num of replicas, etc is a good point. Originally I had wanted to have a different card for the operator vs contour, but didn't get much support for the idea. 

We could do something like that but make the config crd a reference or something? Worth a discussion on. 

Or those fields are just ignored if not using the operator which feels bad as well.

Reviewer 2: I think we're probably going to end up with two CRDs: the operator Contour/ContourDeployment CRD that describes how Contour should be deployed. This should reference a ContourConfig CRD, or reuse the same structs as the ContourConfig CRD, which the controller will watch and handle.

Reviewer 2: Actually, now I've thought about it a little more, I think that using the ContourConfig Spec struct in the Operator's ContourDeployment object is the way to go, so you would end up with something like:
```yaml
kind: ContourDeployment
spec:
  deployment: <replicas and other Contour controller deployment settings>
  config:
    <include ContourConfigSpec struct here>
```

Doing this means that the Operator can simply take whatever is in the `config` stanza, and turn it into the Spec for a ContourConfig object, and that updating ContourConfig automatically updates the ContourDeployment's fields.

Reviewer: Yeah, seems to me like we'll need to have 2 CRDs as well. Will think a little more about embedding the Spec struct vs. using a reference but we can definitely make one of those work.

Author: I'd prefer to not have the reference, having to manage 100 different objects is way too complex for what we're trying to do. That was the idea behind having HTTPProxy be able to work as a single resource, or many. 

To make it simpler, we could put all the CRDs in the Contour repo, and the operator could just reference those CRDs, so it's just a consumer of the `ContourDeployment` (or whatever name it gets).

Reviewer 3: would be good to enumerate what the operator only information is like mentioned above (# of replicas etc.) and see what we need as fields and what we can achieve by convention (eg. namespace of created Contour could be the same the Config CRD is in etc.)

Reviewer 2: Agreed on that last point, @sunjayBhatia, and agreed as well that putting all the CRDs in one place makes sense (probably the Contour repo, but at some point it may be worth extracting them out into a separate repo).

Reviewer 2: The reason I suggested embedding also is because of things like Deployment and ReplicaSet already using that strategy.

#### managed: false

If envoy.managed is false, then Contour won't manage the Envoy fleet/service. It's up to the cluster operator to deploy and manage these resources.
The external ports of the load balancer must match the ports Contour expects (otherwise insecure-secure redirects won’t work).

The external ports should almost always be 80/443.

If Gateway.ControllerName is not set, then the envoy-service-xxx flags passed to Contour will be used to define the ports which Envoy/Service will utilize.

See: [https://github.com/projectcontour/contour/issues/3616](https://github.com/projectcontour/contour/issues/3616)

### Leader Election

Contour will take into account if it is the Leader or not which will represent if it should take action on managing resources or not.

### Finalizers

Finailzers will need to be put into place for the resources which Contour owns (i.e. Envoy) as well a GatewayAPI bits if applicable:

* Envoy Service/Daemonset
* Gateway
* GatewayClass

#### Finalizer workflows:

* If the Envoy service is deleted what happens?
* If the Envoy Daemonset is deleted what happens?
* If the Gateway is deleted what happens?
* If the GatewayClass is deleted what happens?

> ### Exposing Envoy

Reviewer 3: had a thought about this and what the Gateway listener ports "mean" and how we could service multiple scenarios of deployments

how much do users *need to customize* the ports Envoy listens on in the container/how much do we have to think about how the Envoy pods are routable?

If we choose a certain direction, are we locking people who use a managed Envoy to do a specific thing, which may be incompatible with how they have their cluster set up (routable pods or not, CNI, other constraints)?

if the ports on a Gateway listener represent the ports that a user outside a cluster will access apps via and we choose Envoy internal ports for you/make you configure them outside of the Gateway, that seems unwieldy potentially?

also, if we change how we treat those ports (are they on the LB service or just what envoy listens on) halfway through implementation/migration that will probably not be a great experience

thinking about this a bit and seeing all the places we have port configurations etc. in our example deployments (that people right now have the ability to customize to make work) makes me think we need to start earlier with the "gateways everywhere" approach for managed Envoy, and reconcile 2 different types of Gateways, one for configuring how Envoy listens, and another for how Envoy is exposed
- a migration step here could be to keep the Envoy LoadBalancer service and associated implementation logic a bit separate from the Gateway stuff, and make that for users using separate configuration
- another approach would be to just include it in example YAMLs and leave it up to the user how they want traffic into Envoy

Reviewer 3: this was discussed a bit in the doc, but thought I would bring it up again, I dont know if we fully discussed the various migration steps to the end state of things and how all these ports interact when youre not using Gateway API though

Reviewer 3: just worried we get locked into one deployment model or cluster configuration and find it hard to adapt later b/c theres no space in the API for adding more information to a Gateway for example

Reviewer 2: I think it's not stated very clearly, but the assumption is that the Gateway's ports are the ports that the traffic should be sent to from the outside of the Gateway, and that doing other stuff here is implementation detail. I don't think that having the Gateway ports represent the actual container ports Envoy is running on is a good idea without a really good reason.

I think that for this bit, we should start with the easy case, and build outward.

The ports declared on Contour's Gateway are the externally-visible ports that traffic should be sent to via a client. When we are managing the Envoy inbound traffic using a Service or second Gateway object, those should be the requested ports, and we should use a port-translation mechanism like `targetPort` in Service to translate to a high port on the Envoy deployment. As we add more complicated Service requests in, we need to keep this in mind, that the *outer* ports of the Service need to be the same as the ones on the Contour Gateway.

People can customise those ports right now in our example YAMLs, and if they want to do that, managed Envoy is not ready for them yet. I think this is okay to start with. Then, people can tell us what they need, and we can help make that work.

The fact we have to build that is an unavoidable cost of adding the Envoy management functionality in to Contour, sadly.

It's really important to get this right because otherwise we'll break the insecure->secure redirects as well.

Reviewer 2: tl;dr I think we should go ahead with just LB Service to start with, and consider the container ports and mappings an implementation detail.

Author: Feels like a missing bit in the gateway design since an ingress controller doesn't really manage the ports, the external gateway does. 

An ingress gateway us just a cluster service at this point.

Author: I'll update the design. Can start with type loabalancer, but Im going to add a spot to reference a gateway instead. We need to do that, seems silly to not enable gateways when implementing gateways. 🙃

The details defined in the Contour configuration CRD will be used to pre-populate annotations that will be stored in a Kubernetes configmap and referenced from the GatewayClass.Spec.ParametersRef.
The Gateway Stub controller will be utilized to configure a new Kubernetes Service with the annotations defined in the configmap on the GatewayClass and send traffic to Envoy.
Once implementers catch up, specific implementations will be utilized and the stub controller can be deprecated.

New fields will be added to the Contour configuration file which will allow users to define the GVK of a resource used in the GatewayClass along with the ControllerName to be used.
The Operator will still auto-create the GatewayClass/Gateway which will allow the operator to still maintain the proper listeners matching the Gateway Contour is processing, but allow for custom implementations without the need for code changes to Contour.

![alt_text](images/gatewayapi-external-gateway.jpg)

> ### WorkFlows

Reviewer: One thing that's not described here is upgrades. Presumably the flow is - upgrade the Contour deployment, which then notices that its corresponding Envoy is out of date, and updates it? Also, do we care about downgrades / if so, do they work the same way?

If so, worth noting that this now enforces that Contour always gets upgraded before Envoy (which has been my preferred approach to upgrades, but now it's a hard requirement). That means new versions of Contour shouldn't start using new Envoy features on the critical path until they've been out in Envoy for at least a release.

Author: Yup I agree with what you mention @skriss. I can update the design to call that out specifically. 

I think there are also ways that we can support new Envoy features, but pick apart the Envoy version connecting. We'd need to dig into that more, but would be a neat think to at least think about for future work. I know that @sunjayBhatia had some ideas around this as well.

Author: Added a note about upgrades calling this out.

**Assertion**:_ If users are deploying Contour with the Operator, then they are looking for a fully managed environment, meaning, instances of Contour are managed by the Operator, and instances of Envoy are managed by Contour._

When users want to deploy Contour without the Operator, then they have the ability to choose if Envoy should be managed dynamically by Contour or not.

**Assertion**:_ A Contour Configuration CRD is required for Contour to function in all scenarios._

**Assertion**:_ The GatewayClass.Spec.ParamsRef will not be defined since all the work to define how a Contour/Envoy is configured is defined in the Contour configuration CRD._

#### Operator with GatewayAPI (Envoy is managed by convention)

In this mode, the Operator is managing instances of Contour. Each instance of Contour manages its own Envoy fleet by convention (since the Operator is in use). Contour will expose the Envoy fleet by managing a Gateway/GatewayClass/TCPRoute to get traffic to the Envoy instances. The Contour configuration CRD will define how that traffic will get routed or if another Gateway should be managing the traffic (i.e. custom).

> 1. Deploy Operator to Cluster

Reviewer 3: nit: I think some of the markdown rendering on the lists in this and following sections might need a look just b/c of indenting etc. to make sure everything is how you expect it

Author: Yeah I'll clean that up. I used a plugin to move from gDocs to markdown and it didn't do a great job in some places.

That functionality can be described in the Gateway API by HTTPRoutes and TLSRoutes, as they describe Layer 7 ingress, as Contour does today.

Other types of routes include TCPRoutes and UDPRoutes, which are intended for Layer 4 load balancers.
Implementations may also define their own Route objects, at the cost of interoperability.

In terms of its stated goals, Contour is aiming at being an ingress controller - that is, a Layer 7 proxy with some api gateway functions.
Currently, Contour provides "TCP Proxying" that allows the forwarding of TLS streams based on the SNI information, which is precisely what the Gateway API TLSRoute object is for.
If Project Contour (the organisation) does add support for TCP and UDP forwarding, it will not be in the `projectcontour/contour` repo, but will be a separate repo.

This design is intended to cover the initial alpha releases of the Gateway API. 
We will aim to implement the core featureset of the APIs at that point.
We will then work with the upstream community on features that Contour and HTTPProxy currently support, but the Gateway API do not, and how best to represent those features in the APIs.

> So, there are some features that are currently gaps for Contour (wildcard domain names, and exact path matching, for example),

Reviewer 3: nit: Aren't both of these supported now?

Author: yup!

## Non Goals

> - No TCPRoute or UDPRoute support, that is, no support for arbitrary TCP/UDP forwarding.

Reviewer 4: Is there a plan on supporting this in the future?

Author: @xaleeks @youngnick ^^

### Contour Gateway model

Contour considers a Gateway to describe a single Envoy deployment.

> Contour expects to be supplied with a `ControllerName` which is required to match a `GatewayClass`.

Reviewer: Contour's expectation is that there's only one `GatewayClass` for a given `ControllerName`, and only one `Gateway` for that `GatewayClass`, right? Might be helpful to state that explicitly here since that's different than what the API itself allows.

Reviewer 2: Agreed, we should say that Contour will only use a single Gateway and GatewayClass, and say that the rules for choosing a winner in the event there's more than one are specified in this document.

Author: 👍

> Contour expects to be supplied with the full name and namespace of a single Gateway as well as the GatewayClass controller name.

Reviewer 3: I forget the conversation on this, are we still going to require the gateway name/namespace or are we phasing that out for just the controller name? I vaguely remember something about how we choose which Gateway to admit following the conflict resolution guidelines in the spec etc. but I don't remember where that got written down

Reviewer 5: > I forget the conversation on this, are we still going to require the gateway name/namespace or are we phasing that out for just the controller name?

It was my understanding that the gateway ns/name config will be removed and Contour will reconcile all gateway's that reference a gatewayclass matching its `controllerName`.

Author: But there's no way to specify a single controller in the gatewayclass. What happens when multiple gateways match the same controller? 

```
$ kubectl get gatewayclass,gateway -A                                                                                 
NAME                                                  CONTROLLER                             AGE
gatewayclass.networking.x-k8s.io/contour-class        projectcontour.io/ingress-controller   2m47s
gatewayclass.networking.x-k8s.io/contour-class-dupe   projectcontour.io/ingress-controller   2m47s

NAMESPACE        NAME                                       CLASS                AGE
projectcontour   gateway.networking.x-k8s.io/contour        contour-class        2m24s
projectcontour   gateway.networking.x-k8s.io/contour-dupe   contour-class-dupe   20s
```

Author: I guess the suggestion is that the controllername is a convention matching the contour namespace/name. We can do that, just don't think folks are going to get this right.

Contour will update the status of both the Gateway & GatewayClass.

> When a Gateway is not supplied, Contour will not do Gateway API processing.

Reviewer 2: If we are moving to pass a `controllerName` instead of a Gateway name/namespace, we'll need to update this too:
```suggestion
When a GatewayClass `controllerName` is not configured, Contour will not do Gateway API processing.
```

Contour will merge Listeners within its Gateway. (See the detailed design for the exact rules Contour will use for this.)

### Dynamic Envoy Configuration

> Typically, the fleet of Envoy's that connect to Contour to get their configuration information has been managed manually by the cluster operator.

Reviewer 2: Nit:
```suggestion
Typically, the fleet of Envoys that connect to Contour to get their configuration information has been managed manually by the cluster operator.
```

Reviewer 2: Uh, this one isn't changed?

Given a more dynamic configuration API, it's convenient to have the Envoy fleet configuration in Kubernetes be managed by a controller.
This allows cluster operator to simply deploy a Contour deployment, leaving the details about how Envoy is provisioned in the cluster as well as exposed outside of the cluster to a configuration file or CRD.

> This behavior is meant to be optional and layered, such that users who want a fully managed environment can deploy the Contour Operator which manages instances of Contour which in turn manages instances of Envoy.

Reviewer 3: nice 👍🏽

However, other users may not desire that level of controller driven infrastructure.

> This design leaves users with their choice of deployment models. 

Reviewer: I think it'd be helpful to have a detailed design section that goes into this more since this is a big new feature. How does a user opt into getting the Envoy fleet automatically managed? How do they opt out? What are the configuration knobs/where are they specified? etc.

When it calculates the Envoy configuration using the DAG, Contour layers different types of configuration in order.
Currently, Ingress is overwritten by HTTPProxy. 
If an Ingress and a HTTPProxy specify the exact same route, the HTTPProxy will win.

Once we have the Gateway API available as well, the processing order will be Ingress is overwritten by HTTPProxy, is overwritten by the Gateway API.
We could potentially reverse HTTPProxy and the Gateway API here, but this acknowledges that the Gateway API are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.

### Status management

Each object in the Gateway API set has its own status definition. 
In general, Contour will update the status of any object that comes into scope with its details, ensuring that the `observedGeneration` field is set correctly.
When objects fall out of scope, any status set by Contour will not be removed.
It's expected that things that check the status will also check the `observedGeneration` to check if the status information is up-to-date with the current object generation.

### Per-object design notes

#### GatewayClass

> Contour watches for GatewayClass resources which can, optionally, define how an instance of Contour should be configured.

Reviewer 5: We may want to consider supporting a gatewayclass referencing the Contour ConfigMap for params until a config CRD is established.

> Contour watches for GatewayClass resources which can, optionally, define how an instance of Contour should be configured.

Reviewer 2: If we move to only taking GatewayClass, we'll need to remove `optionally` here.

Author: The way @danehans designed it in the operator is you don't need to specify a paramsRef resource which would give you a set of generic defaults.

Reviewer 2: Ah right, I read that wrong. Maybe something like this?
```suggestion
Contour watches for GatewayClass resources which can optionally refer to a configuration object that will then be used for Contour configuration.
```

Author: 👍

> A GatewayClass can reference a configuration resource (Configmap or CRD) which describes how the fleet of Envoys should be provisioned in the cluster along with how they are exposed outside of the cluster.

Reviewer 5: We may want to consider supporting a gatewayclass referencing the Contour ConfigMap for params until a config CRD is established.

Author: yup its left open right now: `A GatewayClass can reference a configuration resource (Configmap or CRD)`

If the GatewayClass doesn't reference any configuration resources, then a standard set of configuration is utilized much like the Contour QuickStart today.

> Contour will need to be configured with the controller name matching the GatewayClass's `Spec.Controller` field.

Reviewer 3: maybe add a note about the convention/format of controller name/default?

Reviewer 2: Agreed that the default should be documented here.

Author: There is no default from what I understand. The examples do generate a field that has a "default", however, it's not applied automatically from code.

Author: @danehans do you foresee a default in Contour?

Reviewer 2: Since we're using the presence of a value in this field to determine whether to switch on the Gateway API processors, it makes sense there's no default specified here.

Reviewer 3: should we instead document the convention of how the controller name string is constructed? I remember someone mentioning we should do `projectcontour.io/<namespace-of-contour-instance>/<some-name>`

Reviewer 3: ah i see a comment about this below, nvm

Author: I actually just took that specific bit out, it doesn't need to match a convention, only be unique.

#### Gateway

> Contour will watch its configured Gateway object.

Reviewer: Since Contour's not configured with a `Gateway` anymore I'd spell this out a little more - i.e. Contour watches the `Gateway` that uses the `GatewayClass` that has the configured `ControllerName`.

Reviewer 2: Agreed, needs a caveat about confict resolution rules too.

Author: There's a note about conflict resolution right after that paragraph.

Author: 👍

## Detailed Design

### Configuration
Contour will have an entry added to the config file for the `ControllerName` it should watch.

> The `ControllerName` takes the format of `projectcontour.io/<namespace>/contour` which requires the `<namespace>` param to be specified matching the namespace where Contour is deployed. 

Author: Since we're going to add Class logic to pick the oldest? The format of this probably doesn't matter, but should follow the upstream convention. 

Should also look to make sure the upstream admission controller also enforces the correct pattern.

Reviewer 2: We can probably say something like "The `ControllerName` should take the format of`, since upstream _recommends_ that controllers namespace this, but does not _require_ it.

This will look like this:
```yaml
gateway:
  controllerName: 
  name: gatewayname  # <--- deprecate
  namespace: gatewaynamespace  # <-- deprecate
```

> Contour will look for a Gateway + GatewayClass that matches the `ControllerName` field specified in the configuration file.

Reviewer 2: This probably needs to be broken up to be more specific, so that we say that Contour will find a single valid GatewayClass, and then any Gateways that reference the GatewayClass.

Similarly, below should say that if Contour doesn't find any matching GatewayClasses, then no Gateways will be selected until there is one.

Author: Yup it states that right below that sentence:
```
If it does not find a matching Gateway/GatewayClass, then it will begin processing GatewayAPI resources once they are available. 
```

Reviewer 2: oh, 🤦 . Thanks.

If it does not find a matching Gateway/GatewayClass, then it will begin processing GatewayAPI resources once they are available. 

> Should two different Gateway or GatewayClasses be found matching the same `ControllerName`, then Contour will utilize the oldest created object and set a status on the other(s) which describes that object as being rejected.

Reviewer 2: This should specify what Condition will be emitted, what Reason should be set, and so on.

If the gateway is removed while Contour is operating, then an error will be logged, and all config associated with the Gateway API will be removed from Envoy.

_NOTE: The fields `name` & `namespace` will be deprecated from the configuration file._

#### Managed Envoy

Contour can optionally provision Envoy infrastructure in a Kubernetes cluster.
A new field will be added to the Contour configuration file which will determine if Contour should dynamically provision a fleet of Envoys with a corresponding service.
The field will be named `envoy` and contain a single value for now of `managed: true` or `managed: false`.

This field will default to `false` if unspecified.

Example config file:
```yaml
envoy:
  managed: true
gateway:
  controllerName: projectcontour.io/ingress-controller
```

_NOTE: Typically Envoy is deployed as a Daemonset, but not required (could also be a Deployment), so the term `fleet` is used to identify the set of Kubernetes managed pods which represent an Envoy instance._

##### managed: true

> If envoy.managed is true, then Contour will automatically create a Envoy fleet/service.

Reviewer 2: Given the discussion of Loadbalancer Services below, I'd suggest something like this:
```suggestion
If envoy.managed is true, then Contour will automatically create a Envoy fleet and a load balancer to get traffic to it. See below for a discussion on options for how Contour will do that.
```

> Additionally, if a `Gateway.ControllerName` is set then the ports that are assigned to the Envoy service will match the Gateway discovered based upon the `ControllerName`.

Reviewer 2: We need to specify what happens in the case that `envoy.managed` is `true` and `gateway.controllerName` is not specified in this section too. Seems likely that we would do the same as below, and the `--envoy-service-*` flags would set the ports.

Additional RBAC permissions will be required to the namespace where Contour is deployed to allow it to dynamically manage Daemonsets/Deployments/Services, etc.

##### managed: false

If envoy.managed is false, then Contour won't manage the Envoy fleet/service.

> It's up to the cluster operator to deploy and managed these resources.

Reviewer 2: We need some specification here of what the cluster operator will need to do. We do this already in the example deployments, but we need to be specific about things like:
- External ports of the load balancer must match the ports Contour expects (otherwise insecure-secure redirects won't work).
- The external ports should almost always be 80 and 443.
- and so on.

Also, we should specify what happens when `managed` is `false` and cluster admins *do* specify Gateway API details.

Initially Contour can do the same work to move that same logic and expose Envoy via annotations on the Service Type=LoadBalancer.

> Additionally, Contour will support utilizing GatewayAPI itself fully and instead of exposing a service with type=LoadBalancer, Contour can create a GatewayClass, Gateway which selects a TCPRoute used to send traffic to the Envoy fleet.

Reviewer 2: Nit: "Eventually" is probably closer to what we mean here, since we're not going to do this right away, I think.

Author: Yup, it's just down to timing I think. I'd like to have the config option in the configuration file. If that's set then Contour doesn't manage the envoy service of type=LoadBalancer, just a regular ClusterIP service.

_[stevesloka] I have some ideas around this that would allow us to manage multiple gateways potentially from a single envoy cleanly. But I don’t want to distract from the main goal of this design. _

> =

Reviewer 2: Nit: Not required?
